JIT: Remove "promoted struct death vars" map

jakobbotsch opened the pull request:
Repurpose the multireg var death flags as general purpose "field death" flags instead. Update the TreeLifeUpdater to take this into account. Update dumping of promoted locals to always display last uses of fields. 

Couldn't reopen my old PR because of force pushing.

jakobbotsch committed:
JIT: Remove "promoted struct death vars" map

Repurpose the multireg var death flags as general purpose "field death"
flags instead.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into promoted-struct-field-deaths

jakobbotsch committed:
Unify and streamline

jakobbotsch committed:
Streamline TreeLifeUpdater

-1% TP on win-x64. Will probably pull this change out.

jakobbotsch committed:
Rename

jakobbotsch commented:
This is blocked on #80611 for now.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into promoted-struct-field-deaths

jakobbotsch committed:
Fix after merge

jakobbotsch committed:
Expand dumping, make the logic of some checks easier

jakobbotsch committed:
Avoid computing previouslyLive when unnecessary

jakobbotsch committed:
Dump field last uses for all promoted locals

jakobbotsch committed:
Avoid unnecessary bit op

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Remove another unnecessary bit op

jakobbotsch commented:
cc @dotnet/jit-contrib PTAL @BruceForstall 

[Only a couple of diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=142223&view=ms.vss-build-web.run-extensions-tab). The new logic allows marking dead fields as dead when a promoted struct is being born, which the old logic did not.
This rarely allows some new copy props (only rarely because the dead assignments would usually be eliminated anyway, except when they are to a promoted struct that has significant padding and holes).

jakobbotsch commented on src/coreclr/jit/morphblock.cpp:
```diff
@@ -1272,7 +1272,12 @@ GenTree* MorphCopyBlockHelper::CopyFieldByField()
+
+            if (m_dstLclNode->IsLastUse(i))
+            {
+                dstFld->gtFlags |= GTF_VAR_DEATH;
```
We can avoid creating IR for these stores at all (#80498). I'll handle that in a follow-up.

BruceForstall commented on src/coreclr/jit/forwardsub.cpp:
```diff
@@ -360,14 +360,17 @@ class ForwardSubVisitor final : public GenTreeVisitor<ForwardSubVisitor>
         {
-            return false;
+            assert(!dsc->lvTracked);
+            GenTreeFlags allFieldsDying =
+                static_cast<GenTreeFlags>(((1 << dsc->lvFieldCnt) - 1) << FIELD_LAST_USE_SHIFT);
```
Looks like this code pattern exists elsewhere, as well (e.g., fgMorphExpandImplicitByRefArg)

jakobbotsch replied:
I added `LclVarDsc::AllFieldDeathFlags()` (for promoted locals) and `LclVarDsc::FullDeathFlags()` (for any local)

BruceForstall reviewed (approved):
LGTM. One minor suggestion. It's awesome to get rid of this map!

jakobbotsch committed:
Address feedback

BruceForstall commented:
I like the cleanup with the new functions, but it looks like your new assert is hitting:
```
Assertion failed 'lvPromoted && (lvFieldCnt <= 4)' in 'System.Decimal:Ceiling(System.Decimal):System.Decimal' during 'Morph - Global' (IL size 36; hash 0xb0841e47; FullOpts)
```

btw, any thoughts on what should happen when (if?) we generalize struct promotion to more than 4 fields, or not all the fields?

jakobbotsch commented:
> I like the cleanup with the new functions, but it looks like your new assert is hitting:
> 
> ```
> Assertion failed 'lvPromoted && (lvFieldCnt <= 4)' in 'System.Decimal:Ceiling(System.Decimal):System.Decimal' during 'Morph - Global' (IL size 36; hash 0xb0841e47; FullOpts)
> ```

Yep, I'll take a look next week. Maybe some of the state is desynced for promoted implicit byrefs while they are being morphed.

> btw, any thoughts on what should happen when (if?) we generalize struct promotion to more than 4 fields, or not all the fields?

The plan is that the generalized promotion pass works only (mostly) by inserting new IR and does not need the special treatment in downstream passes that promoted structs and their fields have today.

jakobbotsch committed:
Fix treatment of fields of implicit byref params with undone promotion

We can get here for locals representing fields of promoted implicit
byrefs that had promotion undone. In those cases we would sometimes
still propagate the last-use information. That's ok in the context of
the optimization we do with the information but it is not technically
correct since another future field access would also be transformed to a
use of the parent local.

jakobbotsch commented:
> Maybe some of the state is desynced for promoted implicit byrefs while they are being morphed.

Nope, the problem is that the we can get here for fields of promoted implicit byrefs too. The logic was technically incorrect there (but ok for the optimization we do with the information). Fixed it, will probably result in a few minor diffs.

jakobbotsch committed:
Restructure

jakobbotsch committed:
Fix typo
